core/src/bytecode/ast/compat.rs

```diff
@@ -1205,10 +1234,31 @@ impl<'ast> FromAst<Node<'ast>> for term::Term {
 
                 Term::StrChunks(chunks)
             }
-            Node::Fun { arg, body } => match arg.data {
-                PatternData::Any(id) => Term::Fun(id, body.to_mainline()),
-                _ => Term::FunPattern((*arg).to_mainline(), body.to_mainline()),
-            },
+            Node::Fun { args, body } => {
+                let body_pos = body.pos;
+
+                // We transform a n-ary function representation to a chain of nested unary
+                // functions, the latter being the representation used in the mainline AST.
+                args.iter()
+                    .rev()
+                    .fold(term::RichTerm::from_ast(body), |acc, arg| {
+                        let term = match arg.data {
+                            PatternData::Any(id) => Term::Fun(id, acc),
+                            _ => term::Term::FunPattern((*arg).to_mainline(), acc),
+                        };
+
+                        // [^nary-constructors-unrolling]: this case is a bit annoying: we need to
+                        // extract the position of the intermediate created functions to satisfy
+                        // the old AST structure, but this information isn't available directly.
+                        //
+                        // What we do here is to fuse the span of the term being built and the one
+                        // of the current argument, which should be a reasonable approximation (if
+                        // not exactly the same thing).
```

Comment:
I'm not sure if it's worth an extra fold, but I think a better position (and a closer approximation of the old behavior) would be to set each "inner" function to have the *entire* fused args+body as its position.

Reply:
I don't have a strong opinion. I think fusing the arguments as done in this PR is a bit exotic at first sight, but has the advantage of being more precise (like for type errors and such, it show you which part of the multi-ary function precisely is at fault, while the old behavior would always show you the whole function for several different terms). On the other hand, matching the old behavior is maybe the most conservative approach.

Reply:
After a second thought, we won't have this issue anymore (or at least not in the same form) once we have only one AST, so I'll proceed with the current version - it's temporary anyways.